### PR TITLE
Sp3 us2

### PR DESCRIPTION
**Pull request by nkalscheuer**
As a sailor, I want real-time navigation updates.
a. *Navigator class takes in start, end, and wind and puts it into the algorithm
b. Event system using streams
c. Tack now calculation to figure out when a sailor should tack
d. Tack now popup UI
e. Off course calculation to trigger a recalculation of the optimal path
f. Ideal heading calculation
g. Ideal heading UI in binnacle

Orig: Got the navigation provider working. Changed ideal heading ui widget as it wasn't in the right place. Also moved the heading widget into its own file. There's some bugs with the map still and with the notifications, but seems like a rendering issue that we'll have to fix. Just want to get everyone synchronized asap.

### CODE REVIEW

File: binnacle/lib/providers/navigation_provider.dart [hunk 1]
```diff
@@ -1,65 +1,284 @@
...
+import 'package:sail_routing_dart/polar_algs/polar_router.dart';
+import 'package:sail_routing_dart/shared/route_math.dart';
 import 'package:sos/models/position_model.dart';
-import 'package:vector_math/vector_math.dart';
+import 'package:sos/models/ideal_heading_model.dart';
+import 'package:sos/models/wind_model.dart';
+import 'package:sos/enums.dart';
+import 'package:flutter/services.dart' show rootBundle;
 
+const metersPerDegree = 111111;
```
Comment (CaseyHillers) [inline binnacle/lib/providers/navigation_provider.dart]: The units for these should all be documented. Additionally, I think we might want to move these into the class so they can be modified during runtime.

File: binnacle/lib/providers/navigation_provider.dart [hunk 2]
```diff
@@ -1,65 +1,284 @@
...
-    _route = new RouteModel(start: _start, end: _end, wind_radians: 0);
-    // new RouteModel(start: _start, end: _end, wind_radians: currentWind.deg);
+  /// Sets close enough with the meter as units. Will be converted to lat/long degrees
+  /// Args:
+  ///   (double) meters : Close enough for finish or tacks in meters
+  /// Returns:
+  ///   void
+  void setCloseEnough(double meters) {
+    if (meters < 0) {
+      // Don't accept negative numbers
```
Comment (CaseyHillers) [inline binnacle/lib/providers/navigation_provider.dart]: This should throw an exception if it's expecting negative numbers to be passed in. That or we check against the absolute value.

File: binnacle/lib/providers/navigation_provider.dart [hunk 3]
```diff
@@ -1,65 +1,284 @@
...
+    if (_currentCheckPoint == 0) {
+      // Route starting
+      _currentCheckPoint = 1;
+      // Update event bus to trigger the start of the route
+      _updateEventBus(NavigationEventType.start);
+    }
+    if (posVec.distanceTo(_end) < _closeEnough) {
+      // Finish!
+      //Push event
+      //eventBus.add(NavigationEvent(eventType: NavigationEventType.finish));
```
Comment (CaseyHillers) [inline binnacle/lib/providers/navigation_provider.dart]: Remove unused code :)

File: binnacle/lib/providers/navigation_provider.dart [hunk 4]
```diff
@@ -1,65 +1,284 @@
...
+import 'package:sos/models/ideal_heading_model.dart';
+import 'package:sos/models/wind_model.dart';
+import 'package:sos/enums.dart';
+import 'package:flutter/services.dart' show rootBundle;
 
+const metersPerDegree = 111111;
+const defaultMaxOffset = 25.0;
+const defaultCloseEnough = 25.0;
+/// Navigation Provider handles new positions and route calculations.
+/// Every position change checks if a new event needs to be emitted or if the route needs to be recalculated.
```
Comment (CaseyHillers) [inline binnacle/lib/providers/navigation_provider.dart]: I think this should be expanded to include explaining how it works along a route and knows which checkpoint it is trying to get to now.

File: binnacle/lib/repository.dart [hunk 5]
```diff
@@ -159,4 +166,10 @@ class Repository {
   setActiveService(ServiceData serviceData) {
     _providerMap[serviceData.category].changeService(serviceData);
   }
+
+  void _setUpNavigationProvider(BehaviorSubject<PositionModel> positionStream) {
```
Comment (CaseyHillers) [inline binnacle/lib/repository.dart]: Does this need its own function?

File: binnacle/lib/providers/navigation_provider.dart (hunk 1 above)
Comment (nkalscheuer) [reply binnacle/lib/providers/navigation_provider.dart]: metersPerDegree is an absolute conversion constant, so should not be changed. The others are just defaults, but the ultimate values can be changed during run time via setters.

Comment (nkalscheuer) [reply binnacle/lib/providers/navigation_provider.dart]: Just added documentation for getters and setters.

File: binnacle/lib/providers/navigation_provider.dart (hunk 4 above)
Comment (nkalscheuer) [reply binnacle/lib/providers/navigation_provider.dart]: Documentation of these are explained in maxOffset and closeEnough public getters and setters now.

File: binnacle/lib/repository.dart (hunk 5 above)
Comment (nkalscheuer) [reply binnacle/lib/repository.dart]: I removed this thanks for pointing that out lol

File: binnacle/lib/providers/navigation_provider.dart (hunk 3 above)
Comment (nkalscheuer) [reply binnacle/lib/providers/navigation_provider.dart]: Removed it

File: binnacle/lib/providers/navigation_provider.dart (hunk 2 above)
Comment (nkalscheuer) [reply binnacle/lib/providers/navigation_provider.dart]: Added a throw FormatException, meant to do this eventually, but forgot

File: binnacle/lib/providers/navigation_provider.dart [hunk 6]
```diff
@@ -1,65 +1,296 @@
...
+  ///   (NavigationEventType): Event that occurred.
+  /// Returns:
+  ///   Nothing, but updates event bus to listeners.
+  void _updateEventBus(NavigationEventType event) {
+    _currentEvent = NavigationEvent(eventType: event);
+    eventBus.add(_currentEvent);
+
+    if (event == NavigationEventType.start) {
+      print("START EVENT!!!");
+      // TODO: Convert this to a stack implementation
```
Comment (dan-rds) [inline binnacle/lib/providers/navigation_provider.dart]: Convert this to a stack implementation?

File: binnacle/lib/providers/navigation_provider.dart [hunk 7]
```diff
@@ -1,65 +1,296 @@
...
+        wind_radians: degToRad(cardinalTransform(wind.deg)));
+    List<Vector2> pointVectors = List<Vector2>();
+    for (var point in points) {
+      pointVectors.add(Vector2(point[0], point[1]));
+    }
+    rm.intermediate_points = pointVectors;
+    return rm;
+  }
+
+  /// Get direction from a start and an end point
```
Comment (dan-rds) [inline binnacle/lib/providers/navigation_provider.dart]: I love the doc strings!!

File: binnacle/test/unit/providers/providerAssets/test_plot.csv [hunk 8]
```diff
@@ -0,0 +1,24 @@
+twa/tws;6;8;10;12;14;16;20
```
Comment (dan-rds) [inline binnacle/test/unit/providers/providerAssets/test_plot.csv]: Do we need this file?

**Pull request merged**